staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go

```diff
@@ -303,10 +315,18 @@ func (e *Store) ListPredicate(ctx genericapirequest.Context, p storage.Selection
 }
 
 // Create inserts a new item according to the unique key from the object.
-func (e *Store) Create(ctx genericapirequest.Context, obj runtime.Object, includeUninitialized bool) (runtime.Object, error) {
+func (e *Store) Create(ctx genericapirequest.Context, obj runtime.Object, createValidation ValidateObjectFunc, includeUninitialized bool) (runtime.Object, error) {
 	if err := rest.BeforeCreate(e.CreateStrategy, ctx, obj); err != nil {
 		return nil, err
 	}
+	// at this point we have a fully formed object.  It is time to call the validators that the apiserver
+	// handling chain wants to enforce.
+	if createValidation != nil {
+		if err := createValidation(obj.DeepCopyObject()); err != nil {
```

Comment:
why this deepcopy?

Reply:
also below

Reply:
> why this deepcopy?

We want to make certain that the validating admission plugins being used can't accidentally mutate the value.  Creates are comparatively uncommon (mostly read load), so the performance impact won't be too bad.

Also, we already deepcopy original objects in the update flow.

Reply:
the interface says that mutation is forbidden. So technically we don't need that. But create is uncommon enough, I don't care much here.

Reply:
> the interface says that mutation is forbidden. So technically we don't need that. But create is uncommon enough, I don't care much here.

I made the comment to point complainers to: "your changes are dropped because you shouldn't make changes" and I do the deepcopy because I know people will mess it up and I can trust them (including future me).